### PR TITLE
LDMD: Append default file extension if -of doesn't contain any

### PR DESCRIPTION
**Pull request by kinke**
Resolves issue #2001.

### CODE REVIEW

**Comment by thewilsonator** [comment]
Would be good to have tests for this.

**Comment by kinke** [comment]
I wanted to make sure all agree on this (not just for LDMD, but for LDC too) before adding the PITA tests.

**Comment by thewilsonator** [comment]
Do we not have the ability to do tests for only LDMD? or are you wanting to do this for both and hence waiting?

**Comment by BBasile** [comment]
the tests should verify this:

| command | Posix output | Windows Output |
| - | - | - |
| file.d -of=file | file| file.exe
| file.d -of=file.plugin | file.plugin| file.plugin
| file.d -lib -of=file.plugin | file.plugin| file.plugin
| file.d -lib -of=file | file.a| file.lib

**Comment by kinke** [comment]
> Do we not have the ability to do tests for only LDMD?

Of course, e.g., dmd-testsuite uses LDMD exclusively.

> or are you wanting to do this for both and hence waiting?

This PR does it for both, as it's implemented in LDC, regardless of being invoked via LDMD or not. We could also keep the previous behavior of not tampering with the `-of` value (for LDC, not LDMD) in case some other tool relied on that, although I think that's highly unlikely as the static/dynamic libs or Windows executables are barely usable without extension.

**Comment by BBasile** [comment]
The corner case is just when a dot is present in the `-of` argument. Unfortunately I don't speak cpp but I have the feeling that you always add the right extension, which must not be done. The right extension is only added when no extension is already present (row 2 & 3).

**Comment by BBasile** [comment]
Also a question: You seem make the change in a file that's not specific to LDMD. Does the fix only applies to LDMD ? (it should not affect LDC2).

**Comment by kinke** [comment]
Don't worry and don't rely on your feelings :], the extension is only added if the filename part doesn't contain a dot (see PR title...). DMD uses the same logic, and `FileName::defaultExt()` is implemented in the DMD front-end (https://github.com/ldc-developers/ldc/blob/master/ddmd/root/filename.d#L409).

**Comment by kinke** [comment]
> Also a question

I already explained 2 posts further up...

**Comment by JohanEngelen** [comment]
I think LDC should do what other compilers do: no automatic file extension adding.

**Comment by kinke** [comment]
> I think LDC should do what other compilers do: no automatic file extension adding.

Which ones did you check? If gcc and clang don't add anything, I very much agree.

**Comment by JohanEngelen** [comment]
I checked clang.

File: driver/cl_options.cpp [hunk 1]
```diff
@@ -59,6 +59,9 @@ cl::list<std::string> runargs(
         "Runs the resulting program, passing the remaining arguments to it"),
     cl::Positional, cl::PositionalEatsArgs);
 
+cl::opt<bool> ldmd("ldmd", cl::desc("Invoked by LDMD?"), cl::ZeroOrMore,
```
Comment (JohanEngelen) [inline driver/cl_options.cpp]: could you rename the `ldmd` variable to `invokedByDMD` or something else more descriptive?

**Comment by kinke** [comment]
https://github.com/ldc-developers/ldc/blob/master/driver/main.cpp#L1031 - shouldn't the default `-shared`` output extension be `.dylib` instead of `.so` on OSX (Darwin)?

**Comment by kinke** [comment]
I'd say let's put this and (adapted) #2021 into 1.2 and then release the first alpha asap.

File: runtime/profile-rt/DefineBuildProfileRT.cmake [hunk 2]
```diff
@@ -16,8 +16,6 @@ if (LDC_WITH_PGO)
         if (NOT (LDC_LLVM_VER GREATER 309))
             set(PROFRT_EXTRA_LDFLAGS "Ws2_32.lib")
         endif()
-    else()
-        set(PROFRT_EXTRA_FLAGS "-fPIC -O3")
```
Comment (kinke) [inline runtime/profile-rt/DefineBuildProfileRT.cmake]: Appropriate optimizations (for MSVC too) are already enforced - we rewrite all `CMAKE_C_FLAGS_{DEBUG,MINSIZEREL,RELWITHDEBINFO}` with ..._RELEASE in `runtime/CMakeLists.txt` (as the C parts of druntime/Phobos are always compiled in release too).

**Pull request merged**